torch/testing/_internal/distributed/nn/api/remote_module_test.py

```diff
@@ -522,30 +520,43 @@ def test_send_remote_module_over_the_wire_script(self):
         expected_unpickled_attrs.append("forward_async")
         expected_unpickled_attrs.append("forward")
 
-        # Create a remote module on worker1 and then pass it to worker2 over the RPC layer.
+        with self.assertRaisesRegex(
+            RuntimeError, "Passing a script RemoteModule over RPC is not supported."
+        ):
+            # Create a remote module on worker1 and then pass it to worker2 over the RPC layer.
+            for remote_module in self._create_remote_module_iter(
+                dst_worker1_name, modes=[ModuleCreationMode.MODULE_CTOR_WITH_INTERFACE]
+            ):
+                # Test querying some simple attributes from worker2.
+                attrs = rpc.rpc_sync(
+                    dst_worker2_name, remote_module_attributes, (remote_module,)
+                )
+
+    @dist_utils.dist_init
+    def test_create_remote_module_by_module_rref(self):
+        if self.rank != 0:
+            return
+        dst_worker1_name = dist_utils.worker_name((self.rank + 1) % self.world_size)
+        dst_worker2_name = dist_utils.worker_name((self.rank + 2) % self.world_size)
+
+        # Create a remote module on worker1 and then pass its `module_rref` to worker2 over the RPC layer.
         for remote_module in self._create_remote_module_iter(
-            dst_worker1_name, modes=[ModuleCreationMode.MODULE_CTOR_WITH_INTERFACE]
+            dst_worker1_name, modes=[ModuleCreationMode.MODULE_CTOR]
         ):
-            # Test querying some simple attributes from worker2.
-            attrs = rpc.rpc_sync(
-                dst_worker2_name, remote_module_attributes, (remote_module,)
+            remote_module2 = rpc.rpc_sync(
+                dst_worker2_name,
+                create_remote_module_by_module_rref,
+                (dst_worker2_name, remote_module.get_module_rref()),
```

Comment:
nice! Do we have a way to ensure that only after all module_rrefs go out of scope (decrefs), the actual RemoteModule will become obselete and gc will handle it? Or we don't need to handle this bc it's guaranteed by RRef internal refcounting logic naturally?

Reply:
We don't need to worry about this, because RRef is similar to a shared pointer that has internal ref counting.